types/src/v0/impls/l1.rs

```diff
@@ -405,12 +408,80 @@ impl L1Client {
     /// called again.
     pub async fn shut_down_tasks(&self) {
-        if let Some(update_task) = self.update_task.0.lock().await.take() {
-            update_task.abort();
+        if let Some(mut update_task) = self.update_task.0.lock().await.take() {
+            update_task.abort_all();
         }
     }
 
+    /// Update stake-table cache on `L1Event::NewFinalized`.
+    fn stake_update_loop(&self, address: Address) -> impl Future<Output = ()> {
+        tracing::debug!("spawn stake table update loop");
+        let opt = self.options();
+        let retry_delay = opt.l1_retry_delay;
+        let chunk_size = opt.l1_events_max_block_range;
+        let state = self.state.clone();
+
+        let stake_table_contract =
+            PermissionedStakeTableInstance::new(address, self.provider.clone());
+
+        let mut events = self.receiver.activate_cloned();
+
+        let span = tracing::warn_span!("L1 client stake_tables update");
+        async move {
+            loop {
+                while let Some(event) = events.next().await {
+                    let L1Event::NewFinalized { finalized } = event else {
+                        continue;
+                    };
+
+                    // Latest finalized block or block contract was deployed at or 0.
+                    let start_block = {
+                        let state = state.lock().await;
+                        state
+                            .snapshot
+                            .finalized
+                            .map(|block_info| block_info.number)
+                            .or(state.stake_table_initial_block)
+                            .unwrap_or(0)
+                    };
+
+                    let chunks = ChunkGenerator::new(start_block, finalized.number, chunk_size);
+                    let mut events: Vec<StakersUpdated> = Vec::new();
+                    for Range { start, end } in chunks {
+                        match stake_table_contract
+                            .StakersUpdated_filter()
+                            .from_block(start)
+                            .to_block(end)
+                            .query()
+                            .await
+                        {
+                            Ok(e) => {
+                                for event in e {
+                                    events.push(event.0)
+                                }
+                                break;
+                            }
+                            Err(err) => {
+                                tracing::warn!(start, end, %err, "Stake Table L1Event Error");
+                                sleep(retry_delay).await;
+                            }
+                        }
+                    }
+                    {
+                        let st = StakeTables::from_l1_events(events);
+                        let mut state = state.lock().await;
+                        state.put_stake_tables(finalized.number, st)
```

Comment:
I think we are missing a "wait for finalized block" somewhere here to ensure that our RPC is not lagging behind. If this is taken care of I think a comment explaining how that's done would assure the reader.